optax/_src/loss.py

```diff
@@ -0,0 +1,148 @@
+# Lint as: python3
+# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Standard losses used in optimisation."""
+
+import chex
+import jax
+import jax.numpy as jnp
+from optax._src import utils
+
+
+def l2_loss(
+    predictions: chex.Array,
+    targets: chex.Array,
+) -> chex.Array:
+  """Calculates the L2 loss of predictions wrt targets.
+
+  Note: the 0.5 term is standard in "Pattern Recognition and Machine Learning"
+  by Bishop, but not "The Elements of Statistical Learning" by Tibshirani.
+
+  References:
+    [Chris Bishop, 2006](https://bit.ly/3eeP0ga)
+
+  Args:
+    predictions: a vector of arbitrary shape.
+    targets: a vector of shape compatible with predictions.
+
+  Returns:
+    the squared error loss.
+  """
+  chex.assert_type([predictions, targets], float)
+  return 0.5 * (predictions - targets)**2
+
+
+def huber_loss(
+    predictions: chex.Array,
+    targets: chex.Array,
+    delta: float = 1.) -> chex.Array:
+  """Huber loss, similar to L2 loss close to zero, L1 loss away from zero.
+
+  If gradient descent is applied to the `huber loss`, it is equivalent to
+  clipping gradients of an `l2_loss` to `[-delta, delta]` in the backward pass.
+
+  References:
+    [Huber, 1964](www.projecteuclid.org/download/pdf_1/euclid.aoms/1177703732)
+
+  Args:
+    predictions: a vector of arbitrary shape.
+    targets: a vector of shape compatible with predictions.
+    delta: the bounds for the huber loss transformation, defaults at 1.
+
+  Returns:
+    a vector of same shape of `x`.
+  """
```

Comment:
I appreciate you documenting the code so well 👍 Thanks for the references

Reply:
:)